File: deeplinkdispatch/src/test/java/com/airbnb/deeplinkdispatch/BaseDeepLinkDelegateTest.java
package com.airbnb.deeplinkdispatch;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import org.jetbrains.annotations.NotNull;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class BaseDeepLinkDelegateTest {

  @Test
  public void testDispatchNullActivity() {
    DeepLinkEntry entry = deepLinkEntry("airbnb://foo/{bar}");
    TestDeepLinkDelegate testDelegate = getOneRegistryTestDelegate(Arrays.asList(new DeepLinkEntry[]{entry}), null);
    String message = null;
    try {
      testDelegate.dispatchFrom(null);
    } catch (NullPointerException e) {
      message = e.getMessage();
    }

    assertThat(message).isEqualTo("activity == null");
  }

  @Test
  public void testDispatchNullActivityNullIntent() {
    DeepLinkEntry entry = deepLinkEntry("airbnb://foo/{bar}");
    TestDeepLinkDelegate testDelegate = getOneRegistryTestDelegate(Arrays.asList(new DeepLinkEntry[]{entry}), null);
    String message = null;
    try {
      testDelegate.dispatchFrom(null, null);
    } catch (NullPointerException e) {
      message = e.getMessage();
    }

    assertThat(message).isEqualTo("activity == null");
  }

  @Test
  public void testDispatchNullIntent() {
    DeepLinkEntry entry = deepLinkEntry("airbnb://foo/{bar}");
    TestDeepLinkDelegate testDelegate = getOneRegistryTestDelegate(Arrays.asList(new DeepLinkEntry[]{entry}), null);

    Activity activity = mock(Activity.class);
    when(activity.getIntent())
      .thenReturn(null);

    String message = null;
    try {
      testDelegate.dispatchFrom(activity);
    } catch (NullPointerException e) {
      message = e.getMessage();
    }

    assertThat(message).isEqualTo("sourceIntent == null");
  }

  @Test
  public void testDispatchNonNullActivityNullIntent() {
    DeepLinkEntry entry = deepLinkEntry("airbnb://foo/{bar}");
    TestDeepLinkDelegate testDelegate = getOneRegistryTestDelegate(Arrays.asList(new DeepLinkEntry[]{entry}), null);

    Activity activity = mock(Activity.class);
    when(activity.getIntent())
      .thenReturn(null);

    String message = null;
    try {
      testDelegate.dispatchFrom(activity, null);
    } catch (NullPointerException e) {
      message = e.getMessage();
    }

    assertThat(message).isEqualTo("sourceIntent == null");
  }

  @Test
  public void testCreateResultAllNull() {
    DeepLinkEntry entry = deepLinkEntry("airbnb://foo/{bar}");
    TestDeepLinkDelegate testDelegate = getOneRegistryTestDelegate(Arrays.asList(new DeepLinkEntry[]{entry}), null);

    String message = null;
    try {
      testDelegate.createResult(null, null, null);
    } catch (NullPointerException e) {
      message = e.getMessage();
    }

    assertThat(message).isEqualTo("activity == null");
  }

  @Test
  public void testCreateResultNullIntent() {
    DeepLinkEntry entry = deepLinkEntry("airbnb://foo/{bar}");
    TestDeepLinkDelegate testDelegate = getOneRegistryTestDelegate(Arrays.asList(new DeepLinkEntry[]{entry}), null);

    Activity activity = mock(Activity.class);
    when(activity.getIntent())
      .thenReturn(null);

    String message = null;
    try {
      testDelegate.createResult(activity, null, null);
    } catch (NullPointerException e) {
      message = e.getMessage();
    }

    assertThat(message).isEqualTo("sourceIntent == null");
  }

  @Test
  public void testCreateResultAllNullData() {
    DeepLinkEntry entry = deepLinkEntry("airbnb://foo/{bar}");
    TestDeepLinkDelegate testDelegate = getOneRegistryTestDelegate(Arrays.asList(new DeepLinkEntry[]{entry}), null);

    Intent intent = mock(Intent.class);
    when(intent.getData())
      .thenReturn(null);
    Activity activity = mock(Activity.class);
    when(activity.getIntent())
      .thenReturn(intent);

    DeepLinkResult result = testDelegate.createResult(activity, intent, null);

    assertThat(result).isEqualTo(new DeepLinkResult(
      false, null, "No Uri in given activity's intent.", null, null, null));
  }

  @Test
  public void testErrorHanderWithDuplicateMartch() {
    String deeplinkUrl = "airbnb://foo/{bar}";
    DeepLinkEntry entry = deepLinkEntry(deeplinkUrl);

    Uri uri = mock(Uri.class);
    when(uri.toString())
      .thenReturn(deeplinkUrl);
    Intent intent = mock(Intent.class);
    when(intent.getData())
      .thenReturn(uri);
    Context appContext = mock(Context.class);
    Activity activity = mock(Activity.class);
    when(activity.getIntent())
      .thenReturn(intent);
    when(activity.getApplicationContext())
      .thenReturn(appContext);

    TestErrorHandler errorHandler = new TestErrorHandler();
    TestDeepLinkDelegate testDelegate = getTwoRegistriesTestDelegate(Arrays.asList(new DeepLinkEntry[]{entry}), Arrays.asList(new DeepLinkEntry[]{entry}), errorHandler);
    DeepLinkResult result = testDelegate.dispatchFrom(activity, intent);

    assertThat(errorHandler.duplicatedMatchCalled()).isTrue();
    assertThat(errorHandler.getDuplicatedMatches()).isNotNull();
    assertThat(errorHandler.getDuplicatedMatches().size()).isEqualTo(2);
    assertThat(errorHandler.getDuplicatedMatches().get(0)).isEqualTo(entry);
    assertThat(errorHandler.getDuplicatedMatches().get(1)).isEqualTo(entry);

    assertThat(result.getDeepLinkEntry().equals(entry));
  }

  @Test
  public void testErrorHandlerNotGettingCalled() {
    String deeplinkUrl1 = "airbnb://foo/{bar}";
    String deeplinkUrl2 = "airbnb://bar/{foo}";
    DeepLinkEntry entry1 = deepLinkEntry(deeplinkUrl1);
    DeepLinkEntry entry2 = deepLinkEntry(deeplinkUrl2);

    Uri uri = mock(Uri.class);
    when(uri.toString())
      .thenReturn(deeplinkUrl2);
    Intent intent = mock(Intent.class);
    when(intent.getData())
      .thenReturn(uri);
    Context appContext = mock(Context.class);
    Activity activity = mock(Activity.class);
    when(activity.getIntent())
      .thenReturn(intent);
    when(activity.getApplicationContext())
      .thenReturn(appContext);

    TestErrorHandler errorHandler = new TestErrorHandler();
    TestDeepLinkDelegate testDelegate = getTwoRegistriesTestDelegate(Arrays.asList(new DeepLinkEntry[]{entry1}), Arrays.asList(new DeepLinkEntry[]{entry2}), errorHandler);
    DeepLinkResult result = testDelegate.dispatchFrom(activity, intent);

    assertThat(errorHandler.duplicatedMatchCalled()).isFalse();
    assertThat(result.getDeepLinkEntry().equals(entry2));
  }

  class TestErrorHandler implements ErrorHandler {

    List<DeepLinkEntry> duplicatedMatches = null;
    boolean duplicateMatchCalled = false;

    public List<DeepLinkEntry> getDuplicatedMatches() {
      return duplicatedMatches;
    }

    public boolean duplicatedMatchCalled() {
      return duplicateMatchCalled;
    }

    @Override
    public void duplicateMatch(@NotNull List<DeepLinkEntry> duplicatedMatches) {
      this.duplicatedMatches = duplicatedMatches;
      duplicateMatchCalled = true;
    }
  }

  private static DeepLinkEntry deepLinkEntry(String uri) {
    return new DeepLinkEntry(uri, DeepLinkEntry.Type.CLASS, String.class, null);
  }

  /**
   * Helper method to get a class extending {@link BaseRegistry} acting as the delegate
   * for the
   *
   * @param deepLinkEntries
   * @return
   */
  private static TestDeepLinkRegistry getTestRegistry(List<DeepLinkEntry> deepLinkEntries) {
    return new TestDeepLinkRegistry(deepLinkEntries);
  }

  private static class TestDeepLinkRegistry extends BaseRegistry {
    public TestDeepLinkRegistry(List<DeepLinkEntry> registry) {
      super(registry, getSearchIndex(registry), new String[]{});
    }

    @NotNull
    private static byte[] getSearchIndex(List<DeepLinkEntry> deepLinkEntries) {
      Root trieRoot = new Root();
      for (int i = 0; i < deepLinkEntries.size(); i++) {
        trieRoot.addToTrie(i, DeepLinkUri.parse(deepLinkEntries.get(i).getUriTemplate()), deepLinkEntries.get(i).getActivityClass().toString(), deepLinkEntries.get(i).getMethod());
      }
      return trieRoot.toUByteArray();
    }
  }

  private static TestDeepLinkDelegate getTwoRegistriesTestDelegate(List<DeepLinkEntry> entriesFirstRegistry, List<DeepLinkEntry> entriesSecondRegistry, ErrorHandler errorHandler) {
    return new TestDeepLinkDelegate(Arrays.asList(getTestRegistry(entriesFirstRegistry), getTestRegistry(entriesSecondRegistry)), errorHandler);
  }

  private static TestDeepLinkDelegate getOneRegistryTestDelegate(List<DeepLinkEntry> entries, ErrorHandler errorHandler) {
    return new TestDeepLinkDelegate(Arrays.asList(getTestRegistry(entries)), errorHandler);
  }

  private static class TestDeepLinkDelegate extends BaseDeepLinkDelegate {

    public TestDeepLinkDelegate(List<? extends BaseRegistry> registries, ErrorHandler errorHandler) {
      super(registries, errorHandler);
    }
  }
}

File: deeplinkdispatch/src/test/java/com/airbnb/deeplinkdispatch/BaseDeepLinkDelegateTest.kt
package com.airbnb.deeplinkdispatch

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

@kotlin.ExperimentalUnsignedTypes
class BaseDeepLinkDelegateTest {
    @Test
    fun testFindMethodDeepLinkWrongMethod() {
        val uriTemplate = "airbnb://foo/{bar}"
        val uriString = "airbnb://foo/1"
        val uri = mockk<Uri>()
        every { uri.toString() } returns uriString
        val entry =
            methodDeepLinkEntry(
                uriTemplate,
                TestDeepLinkClassStatic::class.java.name,
                "wrongMethod",
            )
        val activityMock = mockk<Activity>()
        val testDelegate = getOneRegistryTestDelegate(listOf(entry), null)
        val intent = mockk<Intent>(relaxed = true)
        every { intent.data } returns uri
        every { activityMock.intent } returns intent
        val result =
            testDelegate.createResult(activityMock, intent, testDelegate.findEntry(uriTemplate))
        assertThat(result.isSuccessful).isFalse
        assertThat(result.uriString).isEqualTo(uriString)
        assertThat(result.error).isEqualTo("Deep link to non-existent method: wrongMethod")
        assertThat(result.errorThrowable).isInstanceOf(BaseDeepLinkDelegate.DeeplLinkMethodError::class.java)
        assertThat(result.errorThrowable?.cause).isInstanceOf(NoSuchMethodException::class.java)
    }

    @Test
    fun testFindMethodDeepLinkethod() {
        val uriTemplate = "airbnb://foo/{bar}"
        val uriString = "airbnb://foo/1"
        val uri = mockk<Uri>()
        every { uri.toString() } returns uriString
        val entry =
            methodDeepLinkEntry(uriTemplate, TestDeepLinkClassStatic::class.java.name, "testMethod")
        val activityMock = mockk<Activity>(relaxed = true)
        val testDelegate = getOneRegistryTestDelegate(listOf(entry), null)
        val intent = mockk<Intent>(relaxed = true)
        every { intent.data } returns uri
        every { activityMock.intent } returns intent
        val result =
            testDelegate.createResult(activityMock, intent, testDelegate.findEntry(uriTemplate))
        assertThat(result.isSuccessful).isTrue
        assertThat(result.uriString).isEqualTo(uriString)
        assertThat(result.error).isEqualTo("")
        assertThat(result.errorThrowable).isNull()
    }

    @Test
    fun testFindEntry() {
        val entry = activityDeepLinkEntry("airbnb://foo/{bar}")
        val testDelegate = getOneRegistryTestDelegate(listOf(entry), null)
        val foundEntry = testDelegate.findEntry("airbnb://foo/1")
        assertThat(foundEntry).isNotNull
        assertThat(foundEntry?.deeplinkEntry).isEqualTo(entry)
        assertThat(foundEntry?.parameterMap).isEqualTo(
            parameterMap(
                "airbnb://foo/1",
                mapOf("bar" to "1"),
            ),
        )
        assertThat(testDelegate.findEntry("airbnb://bar/1")).isNull()
    }

    @Test
    fun testAllowedPlaceholderValueOrdering() {
        val entry = activityDeepLinkEntry("airbn{scheme(c|b|a)}://foo{host(1|3|2)}/{bar(someValue|someOtherValue)}")
        val testDelegate = getOneRegistryTestDelegate(listOf(entry), null)
        val matchArrayDump =
            testDelegate.registries
                .singleOrNull()
                ?.matchIndex()
                ?.byteArray
                ?.let { String(it) } ?: ""
        assertThat(matchArrayDump).contains("(a|b|c)")
        assertThat(matchArrayDump).contains("(1|2|3)")
        assertThat(matchArrayDump).contains("(someOtherValue|someValue)")
    }

    @Test(expected = NullPointerException::class)
    @Suppress("UNREACHABLE_CODE")
    fun testDispatchNullActivity() {
        val entry = activityDeepLinkEntry("airbnb://foo/{bar}")
        val testDelegate = getOneRegistryTestDelegate(listOf(entry), null)
        testDelegate.dispatchFrom(null!!)
    }

    @Test(expected = NullPointerException::class)
    @Suppress("UNREACHABLE_CODE")
    fun testDispatchNullActivityNullIntent() {
        val entry = activityDeepLinkEntry("airbnb://foo/{bar}")
        val testDelegate = getOneRegistryTestDelegate(listOf(entry), null)
        testDelegate.dispatchFrom(null!!, null!!)
    }

    @Test(expected = NullPointerException::class)
    fun testDispatchNullIntent() {
        val entry = activityDeepLinkEntry("airbnb://foo/{bar}")
        val testDelegate = getOneRegistryTestDelegate(listOf(entry), null)
        val activity = mockk<Activity>()
        every { activity.intent } returns null
        testDelegate.dispatchFrom(activity)
    }

    @Test(expected = NullPointerException::class)
    @Suppress("UNREACHABLE_CODE")
    fun testDispatchNonNullActivityNullIntent() {
        val entry = activityDeepLinkEntry("airbnb://foo/{bar}")
        val testDelegate = getOneRegistryTestDelegate(listOf(entry), null)
        val activity = mockk<Activity>()
        every { activity.intent } returns null
        testDelegate.dispatchFrom(activity, null!!)
    }

    @Test(expected = NullPointerException::class)
    @Suppress("UNREACHABLE_CODE")
    fun testCreateResultAllNull() {
        val entry = activityDeepLinkEntry("airbnb://foo/{bar}")
        val testDelegate = getOneRegistryTestDelegate(listOf(entry), null)
        testDelegate.createResult(null!!, null!!, null)
    }

    @Test(expected = NullPointerException::class)
    @Suppress("UNREACHABLE_CODE")
    fun testCreateResultNullIntent() {
        val entry = activityDeepLinkEntry("airbnb://foo/{bar}")
        val testDelegate = getOneRegistryTestDelegate(listOf(entry), null)
        val activity = mockk<Activity>()
        every { activity.intent } returns null
        testDelegate.createResult(activity, null!!, null)
    }

    @Test
    fun testCreateResultAllNullData() {
        val entry = activityDeepLinkEntry("airbnb://foo/{bar}")
        val testDelegate = getOneRegistryTestDelegate(listOf(entry), null)
        val intent = mockk<Intent>()
        every { intent.data } returns null
        val activity = mockk<Activity>()
        every { activity.intent } returns intent
        val result = testDelegate.createResult(activity, intent, null)
        assertThat(result).isEqualTo(
            DeepLinkResult(
                isSuccessful = false,
                error = "No Uri in given activity's intent.",
                deepLinkHandlerResult = null,
            ),
        )
    }

    @Test
    fun testErrorHandlerWithDuplicateMatch() {
        val deeplinkUrl = "airbnb://foo/{bar}"
        val matchUrl = "airbnb://foo/bar"
        val entry = activityDeepLinkEntry(deeplinkUrl)
        val deeplinkMatchResult = DeepLinkMatchResult(entry, mapOf(DeepLinkUri.parse(matchUrl) to mapOf("bar" to "bar")))
        val uri = mockk<Uri>()
        every { uri.toString() } returns matchUrl
        val intent = mockk<Intent>(relaxed = true)
        every { intent.data } returns uri
        val appContext = mockk<Context>(relaxed = true)
        val activity = mockk<Activity>(relaxed = true)
        every { activity.intent } returns intent
        every { activity.applicationContext } returns appContext
        val errorHandler = DuplicatedMatchTestErrorHandler()
        val testDelegate = getTwoRegistriesTestDelegate(listOf(entry), listOf(entry), errorHandler)
        val (_, _, _, _, match) = testDelegate.dispatchFrom(activity, intent)
        assertThat(errorHandler.duplicateMatchCalled).isTrue
        assertThat(errorHandler.duplicatedMatches).isNotNull
        assertThat(errorHandler.duplicatedMatches!!.size).isEqualTo(2)
        assertThat(errorHandler.duplicatedMatches!![0]).isEqualTo(deeplinkMatchResult)
        assertThat(errorHandler.duplicatedMatches!![1]).isEqualTo(deeplinkMatchResult)
        assertThat(errorHandler.uriString).isEqualTo(matchUrl)
        assertThat(match).isEqualTo(deeplinkMatchResult)
    }

    @Test
    fun testErrorHandlerNotGettingCalled() {
        val deeplinkUrl1 = "airbnb://foo/{bar}"
        val deeplinkUrl2 = "airbnb://bar/{foo}"
        val matchUrl = "airbnb://bar/foo"
        val entry1 = activityDeepLinkEntry(deeplinkUrl1)
        val entry2 = activityDeepLinkEntry(deeplinkUrl2)
        val uri = mockk<Uri>()
        every { uri.toString() } returns matchUrl
        val intent = mockk<Intent>(relaxed = true)
        every { intent.data } returns uri
        val appContext = mockk<Context>(relaxed = true)
        val activity = mockk<Activity>(relaxed = true)
        every { activity.intent } returns intent
        every { activity.applicationContext } returns appContext
        val errorHandler = DuplicatedMatchTestErrorHandler()
        val testDelegate = getTwoRegistriesTestDelegate(listOf(entry1), listOf(entry2), errorHandler)
        val (_, _, _, _, deepLinkEntry) = testDelegate.dispatchFrom(activity, intent)
        assertThat(errorHandler.duplicateMatchCalled).isFalse
        assertThat(deepLinkEntry!!.equals(entry2))
    }

    @Test
    fun testMissingParameterValueReturnsEmptyString() {
        val uriTemplate = "airbnb://foo"
        val uri = mockk<Uri>()
        every { uri.toString() } returns "airbnb://foo?q"
        val entry = methodDeepLinkEntry(uriTemplate, TestDeepLinkClassStatic::class.java.name, "testMethod")
        val activityMock = mockk<Activity>(relaxed = true)
        val testDelegate = getOneRegistryTestDelegate(listOf(entry), null)
        val intent = mockk<Intent>(relaxed = true)
        every { intent.data } returns uri
        every { activityMock.intent } returns intent
        val result = testDelegate.createResult(activityMock, intent, testDelegate.findEntry(uriTemplate))
        assertThat(result.parameters).isEqualTo(mapOf("q" to ""))
    }

    private fun parameterMap(
        url: String,
        parameterMap: Map<String, String>,
    ) = mapOf(DeepLinkUri.parse(url) to parameterMap)

    private class DuplicatedMatchTestErrorHandler : ErrorHandler() {
        var className: String = ""
        var duplicatedMatches: List<DeepLinkMatchResult>? = null
        var duplicateMatchCalled: Boolean = false
        var uriString: String? = null

        override fun duplicateMatch(
            uriString: String,
            duplicatedMatches: List<DeepLinkMatchResult>,
        ) {
            this.uriString = uriString
            this.duplicatedMatches = duplicatedMatches
            duplicateMatchCalled = true
        }
    }

    private class TestDeepLinkDelegate(
        registries: List<BaseRegistry>,
        errorHandler: ErrorHandler?,
    ) : BaseDeepLinkDelegate(registries = registries, errorHandler = errorHandler)

    companion object {
        private fun activityDeepLinkEntry(
            uri: String,
            className: String = Any::class.java.name,
        ): DeepLinkEntry = DeepLinkEntry.ActivityDeeplinkEntry(uri, className)

        private fun methodDeepLinkEntry(
            uri: String,
            className: String,
            methodName: String,
        ): DeepLinkEntry.MethodDeeplinkEntry = DeepLinkEntry.MethodDeeplinkEntry(uri, className, methodName)

        private fun handlerDeepLinkEntry(
            uri: String,
            className: String,
            methodName: String,
        ): DeepLinkEntry.MethodDeeplinkEntry = DeepLinkEntry.MethodDeeplinkEntry(uri, className, methodName)

        private fun getTwoRegistriesTestDelegate(
            entriesFirstRegistry: List<DeepLinkEntry>,
            entriesSecondRegistry: List<DeepLinkEntry>,
            errorHandler: ErrorHandler,
        ): TestDeepLinkDelegate =
            TestDeepLinkDelegate(listOf(testRegistry(entriesFirstRegistry), testRegistry(entriesSecondRegistry)), errorHandler)

        private fun getOneRegistryTestDelegate(
            entries: List<DeepLinkEntry>,
            errorHandler: ErrorHandler?,
        ): TestDeepLinkDelegate = TestDeepLinkDelegate(listOf(testRegistry(entries)), errorHandler)
    }
}

object TestDeepLinkClassStatic {
    @DeepLink("airbnb://foo/{bar}")
    @JvmStatic
    fun testMethod(context: Context): Intent = Intent()
}
